setup-go/action.yaml

```diff
@@ -0,0 +1,46 @@
+# Copyright 2024 Knative Authors
+# SPDX-License-Identifier: Apache-2.0
+
+name: 'Setup Go'
+description: |
+  This action sets up golang with the default version defined by the
+  Knative project maintainers
+
+inputs:
+  go-version:
+    description: |
+      override the go version defined by the Knative project
+
+    required: false
+
+runs:
+  using: "composite"
+  steps:
+    - name: determine go version
+      shell: bash
+      run: |
+        shopt -s extglob
+        GO_SETUP_VERSION="${{ inputs.go-version }}"
+
+        if [[ -z "${GO_SETUP_VERSION}" ]]; then
+          REF_NAME="${{ github.base_ref || github.ref }}"
+          case "${REF_NAME}" in
+            *release-*)
+              GO_SETUP_VERSION="1.21"
+              ;;
+            *main*)
+              GO_SETUP_VERSION="1.22"
+              ;;
+            *)
+              GO_SETUP_VERSION="1.22"
+              ;;
+          esac
+        fi
```

Comment:
Thinking about the release process - I don't think this will match up with prow since we usually cut the branch first here before we trigger the release.

Maybe we just simplify this to be 1.22 and run the latest - in theory certain repos can pin their toolchain now using the toolchain directive